src/Nethermind/Nethermind.Core/Buffers/SpanSource.cs

```diff
@@ -0,0 +1,214 @@
+
+    public SpanSource(byte[] array)
+    {
+        _obj = array;
+    }
+
+    // TODO: if TinyArray used, casting should be done to ISpanSource not CappedArraySource.
+    // public SpanSource(ISpanSource source)
+    // {
+    //     _obj = source;
+    // }
+
+    public SpanSource(CappedArray<byte> capped)
+    {
+        _obj = new CappedArraySource(capped);
+    }
+
+    public static implicit operator SpanSource(byte[] bytes) => new(bytes);
+
+    // TODO: make it correct
+    public int MemorySize
+    {
+        get
+        {
+            const int objSize = MemorySizes.RefSize;
+
+            var obj = _obj;
+
+            if (obj == null)
+                return objSize;
+
+            if (obj is byte[] array)
+            {
+                return ReferenceEquals(array, Empty._obj)
+                    ? objSize
+                    : objSize + MemorySizes.ArrayOverhead + array.Length;
+            }
+
+            return objSize + Unsafe.As<CappedArraySource>(obj).MemorySize;
+        }
+    }
+
+    public int Length
+    {
+        get
+        {
+            var obj = _obj;
+            if (obj is byte[] array)
+                return array.Length;
+
+            return Unsafe.As<CappedArraySource>(obj).Length;
+        }
+    }
+
+    public bool SequenceEqual(ReadOnlySpan<byte> other)
+    {
+        var obj = _obj;
+        if (obj is byte[] array)
+            return array.AsSpan().SequenceEqual(other);
+
+        return Unsafe.As<CappedArraySource>(obj).SequenceEqual(other);
+    }
+
+    public int CommonPrefixLength(ReadOnlySpan<byte> other)
+    {
+        var obj = _obj;
+        if (obj is byte[] array)
+            return array.AsSpan().CommonPrefixLength(other);
+
+        return Unsafe.As<CappedArraySource>(obj).CommonPrefixLength(other);
+    }
+
+    public Span<byte> Span
+    {
+        get
+        {
+            var obj = _obj;
+
+            if (obj is null)
+                return Span<byte>.Empty;
+
+            if (obj is byte[] array)
+                return array.AsSpan();
+
+            return Unsafe.As<CappedArraySource>(obj).Span;
+        }
+    }
+
+    public bool IsNotNull => !IsNull;
+    public bool IsNull => _obj == null;
+    public bool IsNotNullOrEmpty
+    {
+        get
+        {
+            var obj = _obj;
+
+            if (obj == null)
+                return false;
+
+            if (obj is byte[] array)
+                return array.Length != 0;
+
+            return Unsafe.As<CappedArraySource>(obj).Length != 0;
+        }
+    }
+
+    public static readonly SpanSource Empty = new([]);
+
+    public static readonly SpanSource Null = default;
+
+    public bool Equals(SpanSource other)
+    {
+        Span<byte> comparand = other.Span;
+
+        var obj = _obj;
+        if (obj is byte[] array)
+        {
+            return array.AsSpan().SequenceEqual(comparand);
+        }
+
+        return Unsafe.As<CappedArraySource>(obj).SequenceEqual(comparand);
```

Comment:
Question I'd have is `is CappedArraySource` as its a simple sealed class a faster type check than `is byte[]`

Reply:
Let me test it. This `Unsafe.As` was put when the `ITinyArray` was in this PR, while it's no longer the case. As it's no longer the case, maybe with the `sealed` we can get as fast as this Unsafe. Will report.

Reply:
I changed it for Span and `MemorySize`. It seems that it plays with the JIT. For the `MemorySize` it looses one jump in the `CappedArray` size. For the `Span` it no longer inlines the callsite but removes `int 3` for overflow and returns immediately after testing for being the capped array. Pushed already.